### PR TITLE
Route stdout/stderr to different buffers based on regexps

### PR DESCRIPTION
**Pull request by cskksc**
Check out the very short [demo](https://www.youtube.com/watch?v=nhej5XE2THE&feature=youtu.be) (15 sec), with 

`(setq cider-redirect-output-regexps '(("INFO: .*" . "*cider-log*")))`.

If there are multiple matches, we use the first match but display a warning.

-----------------

Before submitting the PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [ ] The commits are consistent with our [contribution guidelines][1]
- [ ] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (`make test`)
- [ ] The new code is not generating bytecode or `M-x checkdoc` warnings
- [x] You've updated the changelog (if adding/changing user-visible functionality)
- [x] You've updated the readme (if adding/changing user-visible functionality)
- [ ] You've updated the refcard (if you made changes to the commands listed there)

Thanks!

[1]: https://github.com/clojure-emacs/cider/blob/master/.github/CONTRIBUTING.md

### CODE REVIEW

File: cider-repl.el [hunk 1]
```diff
@@ -571,12 +579,29 @@ If BOL is non-nil insert at the beginning of line."
             (set-marker cider-repl-output-end (1- (point)))))))
     (cider-repl--show-maximum-output)))
 
+(defun cider--interactive-out-buffer (string default)
```
Comment (cskksc) [inline cider-repl.el]: This is not the best name for this. Suggestions please.

File: cider-repl.el [hunk 2]
```diff
@@ -155,6 +155,14 @@ you'd like to use the default Emacs behavior use
   :group 'cider-repl
   :package-version '(cider . "0.11.0"))
 
+(defcustom cider-redirect-output-regexps
+  '()
+  "A list to define mappings from regexps to buffer names.
```
Comment (bbatsov) [inline cider-repl.el]: I guess it'd be nice to give an example of the mappings in the docstring.

**Comment by dpsutton** [comment]
Are we building a new logging system? I know this issue popped up on someone's project, but is this a poorly implemented logging system? Should the "fix" be to use tools.logging or timbre?

**Comment by dpsutton** [comment]
Good point. I'm just worried someone will pull these projects up in another IDE/repl/whatever and just be deluged by stuff because we are hiding java.lang.System/out

**Comment by cskksc** [comment]
@bbatsov Done.

@dpsutton We'll only redirect the output for the regexps users have configured. But it's true that this may be confusing and unexpected at times. Any ideas ?

**Comment by bbatsov** [comment]
Ideally we should ask some users why are they logging to the console before making a decision. I really wonder why they prefer such an unconventional form of logging...

**Comment by expez** [comment]
A while back I noticed that when I call `start` at the REPL I got logs in the REPL.  But when I relied on `cider-refresh` to call `start` for me I got the project's logs in `nrepl-server`.  I'm not sure why this happened, but whenever I tripped up I got an unusable REPL until I restarted the app.

I use timbre for logging and this probably happened because I had the `println` appender as part of the config (I've since removed this appender altogether).  I think that appender is part of the default configuration.  In any event, using the println appender might be something people do because you can then just start your app with `java -jar ... >> log_file 2>&1`.  I expect this pattern comes more naturally to many, instead of say using a file-appender.

In any event, I'll be glad to get a separate buffer for all my logs :)

**Comment by agzam** [comment]
I feel bad. I think it was me who started complaining about this. When I switched from lein to boot, setting `:verbose` option for cljs-compiler as it turned out works differently. When working with lein everything (verbose log) would appear right in the repl buffer, with boot only warnings and errors (and the rest would go to nrepl-server buffer). It looked strange to me at that time. Now I see that I was wrong. It seems it's better that way.

**Comment by bbatsov** [comment]
I'll leave this PR for a while to collect more feedback.

@cskksc Sorry for asking you to work on this when it might not get merged in the end. After @dpsutton's remark I really started doubting whether this was a good idea or not.

**Comment by bbatsov** [comment]
//cc @cch1

**Comment by cskksc** [comment]
@bbatsov No problem. I agree with @dpsutton that this may cause a lot of confusion.

**Comment by bhagany** [comment]
fwiw, I like the redirect buffers.  While I don't quite understand logging to stdout and not wanting to see the output at the repl, I do understand getting used to that behavior once you discover it.  Anyway, `cider-redirect-output-regexps` is the kind of thing that people won't se unless they look for it, and they'll be happy when they find it.  I think they only time it will cause confusion is if they `setq` and forget, and that's true of everything in emacs.

@agzam you shouldn't feel bad.  People have been complaining about output unexpectedly going to the server buffer for at least a year, and I'd be surprised if it weren't longer (this is pretty much a longstanding nrepl bug that CIDER now works around).  Just in the boot world, this bug caused _all_ output from pods to go to the server buffer, which is how I encountered it.  And anyway, I only found your bug report on boot-cljs after #387 was already merged :).

**Comment by bbatsov** [comment]
My main issue with adding this feature is that it adds a certain overhead to every output operation and it makes the pretty convoluted output handlers even more messy. If there's no compelling reason to support redirecting output then it's probably not worth doing it. On the other hand if the messy nREPL handlers were reworked in some saner manner this might be more doable. 

One other thing to consider is do you really want to redirect output from code that was evaluated in the REPL itself?

**Comment by bhagany** [comment]
Good points.  I don't really see myself using this feature if it's merged - I don't want the mental overhead of remembering which output goes where.  I do have some sympathy for people who were relying on the old behavior, but if tailing a file in eshell or whatever is deemed a good enough workaround, I'm fine with that.

**Comment by bbatsov** [comment]
There are also `auto-revert-tail-mode` and `itail` - I'd say they are more convenient for this purpose than using a shell. :-)

**Comment by cch1** [comment]
Thanks @bbatsov for bringing me into this conversation.

I am using `tools.logging` with slf4j+log4j.  Using a console logger sends output to the STDOUT defined before Clojure starts.  I could log to a file instead of the console and `auto-revert-tail-mode`** but then behavior for non-cider uses (e.g. `lein X` from the command line) would change.  I could manage two loggers (or two profiles) -but now this is getting a convoluted.

I'm pretty happy now using `cider-jack-in` and opening a buffer on `*nrepl-server*`.  At one point I was under the impression that buffer would become unavailable on a future release -that would be painful.  The ability to redirect output to multiple buffers is not a huge deal for me as very little non-log4j-console output ends up on STDOUT.  The one regular exception, ironcially, being nrepl startup ("nREPL server started on port 64458 on host 127.0.0.1 - nrepl://127.0.0.1:64458").  The ability to segregate STDOUT and STDERR (`*nrepl-server-stdout*` and `*nrepl-server-stderr*`) would be nice...

The ability to add targeted hooks around the startup of the `*nrepl-server*` buffer would be cool.  I would hook in `log4j-mode` or `logview-mode` and `read-only-mode`, for example.  As an aside, why is the `*nrepl-server*` not read-only by default?

**in my experience logging to a file and using `auto-revert-tail-mode` is unsatisfactory.  Having a new process reopen the same file for output (e.g. restarting the REPL) causes the mode to stop working.  It's only when you realize that obvious log messages are missing that you realize you need to revisit the file manually.  PITA.

**Comment by bbatsov** [comment]
> I'm pretty happy now using cider-jack-in and opening a buffer on *nrepl-server*. At one point I was under the impression that buffer would become unavailable on a future release -that would be painful.

That just happened. The old behaviour was simply the result of a bug, that was finally fixed on `master`. :-)

**Comment by cch1** [comment]
Uh oh.  So what is the proposed work-around?  If this PR is the workaround, then I'm all for it.  Losing the visibility of STDOUT of the nrepl process will be painful.

**Comment by bhagany** [comment]
@cch1 you won't lose visibility - the output that used to go to the server buffer will now go to your repl buffer, except for the nrepl startup messages you mentioned.

**Comment by dpsutton** [comment]
@cch1 in the intermediate time while we discuss this, it is easy to restore the old behavior by running CIDER from melpa stable which will reference a version of cider-nrepl from before this behavior change.

Also, I want to summarize what your needs are to make sure I understand them, allowing you to correct any misunderstandings.

You don't particularly care where output happens, as long as there is a buffer that has all output. You are OK with stdout happening in the repl.

If the above is correct, I think you should be fine. This change doesn't prevent (to my knowledge) the output to the nrepl server messages buffer. What it does is route output additionally to all interested buffers, ie, repls.

**Comment by bhagany** [comment]
> This change doesn't prevent (to my knowledge) the output to the nrepl server messages buffer. What it does is route output additionally to all interested buffers, ie, repls.

I incorrectly implied that stdout/stderr output would no longer go to the server buffer.  What @dpsutton said here is correct.

**Comment by bbatsov** [comment]
> I incorrectly implied that stdout/stderr output would no longer go to the server buffer. What @dpsutton said here is correct.

Hmm, why would it still go there? I thought your patch redirected all related IO streams. In general I don't think anything should go to the nrepl-server buffer - it exists just because this is how the Emacs process spawning facilities work. Unless some output came straight from leiningen (or whatever was used to start the nrepl-server) it doesn't belong there.

**Comment by bhagany** [comment]
To be honest, I don't really know why output still goes to the server buffer.  When I wrote it, I thought I was completely redirecting the java streams, and it surprised me that it works the way it does.  I didn't investigate much, but I can if it's important to you.

**Comment by bbatsov** [comment]
> I didn't investigate much, but I can if it's important to you.

It's not super important, I just really don't want people to rely on something that's an implementation glitch. As this bug has been around since day one many people thought it was some feature. :-)

**Comment by bhagany** [comment]
Alright, I already have a few theories.  I should have some time this weekend.

**Comment by bhagany** [comment]
Well, it turned out to be the simplest thing: https://github.com/clojure-emacs/cider-nrepl/pull/388

**Comment by dpsutton** [comment]
From @cch1 

>  If this PR is the workaround, then I'm all for it. Losing the visibility of STDOUT of the nrepl process will be painful. 

Do we maybe want to think about that pull request before merging it? I think we may be removing their relied upon behavior for a bugfix that no one is really in a rush for. I agree that its the correct thing to do, but in time and not immediately. thoughts?

**Comment by bhagany** [comment]
I'm fine either way

**Comment by bbatsov** [comment]
> Do we maybe want to think about that pull request before merging it? I think we may be removing their relied upon behavior for a bugfix that no one is really in a rush for. I agree that its the correct thing to do, but in time and not immediately. thoughts?

I'll think about it, but this is definitely going to get merged sooner or later. Postponing this without thinking of some alternative log handling won't serve anyone.

**Comment by cch1** [comment]
I wish this were not viewed as just a hack for log handling.  What other early-loaded Java code binds to STDOUT/STDERR?  They too will be affected.

**Comment by bbatsov** [comment]
> I wish this were not viewed as just a hack for log handling. What other early-loaded Java code binds to STDOUT/STDERR? They too will be affected.

Well, this output is not going to be visible anywhere unless you're using `jack-in`. What happens when using `cider-connect` and there's no process buffer then? That output normally just gets lost. Seems reasonable to just treat all the output in a consistent manner.

**Comment by cch1** [comment]
So the principle is going to be that `cider-jack-in` swallows STDOUT?  Today if I run `cider-jack-in` or `cider-connect`, I am able to see STDOUT (one in emacs, the other from the shell).  Now `cider-jack-in` will be unusable.  :(

**Comment by bhagany** [comment]
> Well, this output is not going to be visible anywhere unless you're using jack-in. What happens when using cider-connect and there's no process buffer then? That output normally just gets lost.

stdout and stderr used to go to whatever process you're `cider-connect`ing to.  This could be a terminal window somewhere, and I think it would only be lost if the process is backgrounded.

**Comment by cch1** [comment]
> stdout and stderr used to go to whatever process you're cider-connecting to. This could be a terminal window somewhere, and I think it would only be lost if the process is backgrounded.

Exactly.

**Comment by bhagany** [comment]
> Now cider-jack-in will be unusable. :(

I don't understand why?

**Comment by cch1** [comment]
@bhagany IF STDOUT is swallowed by the changes recently made on master (per @bbatsov 's comment above) then `cider-jack-in` will be unusable for me.  Assuming your PR can direct STDOUT from the jacked-in nrepl process to a buffer based on a regexp, it would restore the usability

**Comment by bbatsov** [comment]
> stdout and stderr used to go to whatever process you're cider-connecting to. This could be a terminal window somewhere, and I think it would only be lost if the process is backgrounded.

I meant it's not available in Emacs and if you don't redirect it there you won't see it.

**Comment by bhagany** [comment]
> IF STDOUT is swallowed by the changes recently made on master (per @bbatsov 's comment above) then cider-jack-in will be unusable for me.

I'm not sure what comment you're referring to here, but in any case, all of the output goes to the repl buffer.  It's not swallowed.

**Comment by bhagany** [comment]
> I meant it's not available in Emacs and if you don't redirect it there you won't see it.

Ah, okay, got it.

**Comment by expez** [comment]
It seems we're still debating whether this is a good idea or not, and I'd like to raise the point that for certain apps it's a huge problem when you're doing dev work at e.g. `trace` or `debug` presently.  When everything is dumped into the REPL you lose your prompt (due to scrolling or a scrolling related bug) and am unable to interact with the repl.  

Furthermore, any interactions, while your app is trace logging in the background, become potentially unreadable, and you certainly can't interact with the result (e.g. to copy paste) because it will scroll of the screen too fast.

The last time I got into this unfortunate situation I had a hard time recovering my REPL even after stopping the app.  The prompt had disappeared altogether and I was left looking for a point in the buffer where I could hit <kbd>RET</kbd> to restore it.  Is it the last line?  At the end of the last line?  The second to last line?  I can't remember the answer anymore, but I did get my prompt back, eventually.

**Comment by bbatsov** [comment]
It's certain that this feature has certain usefulness. As I said before - my main issue with it is the intrusive implementation right now (mandated by the poor design of the nREPL response handling code). We spoke with @cskksc in private and decide it's best to focus on improving the design of the nREPL handlers before trying to solve the output redirection problem.

**Comment by jcf** [comment]
> It's certain that this feature has certain usefulness. As I said before - my main issue with it is the intrusive implementation right now (mandated by the poor design of the nREPL response handling code). We spoke with @cskksc in private and decide it's best to focus on improving the design of the nREPL handlers before trying to solve the output redirection problem.

@bbatsov can you revert the change in master to bring back the old behaviour of logging to the nREPL server buffer?

The "fix" has broken Emacs for me when log output gets large (I've got a 5.8M REPL buffer that causes Emacs to hang every time I interact with it), and I have to manually clear my buffer losing interesting information every time.

I have log rotation via logback, and can send things to both stdout and files just fine, but with what's in master Emacs is now performing pretty badly.

When nREPL design is improved maybe change the way logging works in Cider?

**Comment by bbatsov** [comment]
> @bbatsov can you revert the change in master to bring back the old behaviour of logging to the nREPL server buffer?

That's not going to happen. I think that the benefits of properly routing the output outweigh the downsides - after all logging to stdout can always be disabled. Fixing the REPLs performance when the text there gets a lot is something else that we should eventually address. 

@cskksc is now working on improvements that should make the stdout redirect idea feasible, so your inconvenience should be temporary. 

> When nREPL design is improved maybe change the way logging works in Cider?

It's unlikely that this is going to happen any time soon. nREPL has seen pretty much no changes in quite a lot of time and I don't see a lot of volunteers to work on improving it.

**Comment by bbatsov** [comment]
Btw, applying this patch locally or going back to CIDER 0.14 is also an option for the time being.

**Comment by cch1** [comment]
> the benefits of properly routing the output outweigh the downsides

What are the benefits?

**Comment by bhagany** [comment]
> the old behaviour of logging to the nREPL server buffer

I'd like to address this, because there seems to be some misapprehension about it.  The issue that's fixed in master isn't about logging.  CIDER doesn't know anything about logging in particular. The fix is about output streams.  To illustrate,

```clojure 
user> (.println java.lang.System/out "foo")
```

If you expect that to print something at the repl, you'd be disappointed if you're using a stable CIDER release.  This is what the fix is aimed at.  CIDER did this with `System/err` as well, like with exceptions,

```clojure
user> (try
  (throw (Exception. "hi"))
  (catch Exception e
    (.printStackTrace e)))
```

The upshot here is that Java stuff couldn't print to your repl before, and that caused lots of problems that people can't fix themselves.  In contrast, redirecting logs is a matter of configuration.

**Comment by cch1** [comment]
@bhagany , when you say "print to the repl", I've never wanted/asked for/seen that behavior. In my experience, `(.println java.lang.System/out "foo")` sends output to my `*nrepl-server*` buffer, not to my *cider-repl* buffer.  

This is what I want and what I fear will be taken away be the changes @bbatsov referenced above.

Maybe I am confused as to what is actually being changed... I hope so.  Without the ability to have java's `java.lang.System/out` in a buffer somewhere, I'm blind.

**Comment by dpsutton** [comment]
@cch1

**Comment by bhagany** [comment]
> when you say "print to the repl", I've never wanted/asked for/seen that behavior

Plenty of others do want it and have asked for it\*, and they are justified in doing so.  You have workarounds for your problem, whereas the people on the other side of this are stuck if it doesn't change.  As for seeing it, try it in `lein repl`, and I would guess, many other repls that people use regularly.

I don't understand why `(.println java.lang.System/out "foo")` having a different behavior from `(println "foo")` should be the default behavior.  It's certainly not intuitive that it should be this way.

> Without the ability to have java's java.lang.System/out in a buffer somewhere, I'm blind.

It seems to me that `*cider-repl*` counts as "a buffer somewhere", but in any case, this PR is being designed to accommodate your use case.  It's being worked on, and you'll be able to put your log messages in the buffer of your choosing, assuming there's a regex that matches them.

\* https://github.com/clojure-emacs/cider/issues/1896
https://github.com/clojure-emacs/cider/issues/1598
https://github.com/clojure-emacs/cider/issues/1404
https://github.com/clojure-emacs/cider-nrepl/issues/381
https://github.com/boot-clj/boot-cljs/issues/138
https://github.com/boot-clj/boot/issues/544

**Comment by cch1** [comment]
> This output goes to the nrepl-server buffer. It now (as of a recent patch) also goes to the cider buffer.

Thanks @dpsutton , for clarifying the issues involved here.  I honestly did not completely understand them before.

The patch behavior seems to introduce redundancy.  What's the point of even having the *nrepl-server* buffer if its content is duplicated in *cider-nrepl*?

When using Cider to connect to programs emitting to STDOUT as a general practice (CLI tools, etc) it seems like the nrepl dialog is going to be conflated with STDOUT.

If the fear is that people aren't looking at `*nrepl-server*` and thus need to be made aware of the their own program's STDOUT, it seems like an easy answer: open the `*nrepl-server*` buffer.   Is that unreasonable?

(Now just reading @bhagany 's comment)  Before I respond, please know that I appreciate everyone's efforts here.  And the prospect of an eventual fix to route based on regex seems like it might address my use case nicely.  Now responding to your particular points:

I will have an unpleasant workaround to my particular problem.  But don't others have an even easier workaround: open the `*nrepl-server*` buffer.  Am I missing something?

When I run `lein repl` in a terminal, I know I only have one output stream (unless I redirect STDOUT to a file).  I expect more from a windowing/multibuffer environment like emacs.

My position: STDOUT is not the same as the nrepl dialog unless I'm in a constrained environment.

**Comment by bhagany** [comment]
> The patch behavior seems to introduce redundancy.

The redundancy is resolved here: https://github.com/clojure-emacs/cider-nrepl/pull/388.

> When using Cider to connect to programs emitting to STDOUT as a general practice (CLI tools, etc) it seems like the nrepl dialog is going to be conflated with STDOUT.

nrepl dialog _is_ stdout.

> Is that unreasonable?

In my opinion, yes.  In the boot ecosystem in particular, the output that is split between buffers is extremely confusing, because parts of it go through a Java intermediary.

> I will have an unpleasant workaround to my particular problem.

Why is it unpleasant?  It's a regex that, as far as I can tell, meets your wants completely.  That's less unpleasant than not working without any options to fix it, which is what you're advocating I deal with for my use case.

**Comment by dpsutton** [comment]
> When I run lein repl in a terminal, I know I only have one output stream (unless I redirect STDOUT to a file). I expect more from a windowing/multibuffer environment like emacs.

That's one way to look at it. Another is to ensure that repl's meet a standard. I would hate to work on a project with someone who had a suppressing repl and I was subjected to lots of messages. This makes me not want to make CIDER's repl suppress things that may bother the rest of the ecosystem. Also why I'm not super sold on the notion of filtering output from the repl buffer, this patch.

**Comment by bbatsov** [comment]
I've said this numerous times and I'm tired of repeating it - the old behavior was always considered a bug by me. Some output got bound early and ended up in a buffer **dedicated exclusively** for the process that spawned the nREPL server - e.g. `lein`, `boot`, whatever. People were warned not to rely on this and yet they did. We're considering the implementation of the redirect feature just for the sake of such users, although in principle I agree with @dpsutton that re-routing output in this manner is probably not a good idea in general.

Btw, we already have a feature to route the output of interactive evaluations (those done in source buffers) to an output buffer instead of to the REPL - probably people unhappy about the recent changes should find it useful. I don't recall whether we had something similar for the REPL itself.

**Comment by bbatsov** [comment]
I guess we can close this for now.

**Comment by cch1** [comment]
> Btw, we already have a feature to route the output of interactive evaluations (those done in source buffers) to an output buffer instead of to the REPL - probably people unhappy about the recent changes should find it useful. I don't recall whether we had something similar for the REPL itself.

I remain frustrated by console logging being intermingled with REPL output.  I realize this ship has already sailed... but can you summarize the current options _other than switching to logging to a file_? 
```
;; CIDER 0.21.0 (New York), nREPL 0.6.0
;; Clojure 1.10.0, Java 1.8.0_162
```

**Comment by bbatsov** [comment]
> I remain frustrated by console logging being intermingled with REPL output. I realize this ship has already sailed... but can you summarize the current options other than switching to logging to a file?

Well, I wouldn't say that ship has sailed - it's just nobody wanted to tackle the flexible routing idea. I still believe that's the best course of action and I hope at some point we'll revisit this.

The old behaviour (where some things would end up in the server's buffer is still available) it's just not the default anymore. Just set `cider-redirect-server-output-to-repl` to nil and you're done.

**Comment by cch1** [comment]
> The old behaviour (where some things would end up in the server's buffer is still available) it's just not the default anymore. Just set `cider-redirect-server-output-to-repl` to nil and you're done.

That worked perfectly.  Thanks!
